### DISCUSSION TITLE
Systemd Needs Delay on startup RPi4B??

### DISCUSSION THREAD

**Discussion by anonyada** [GerryDazoo/Slinger#152] [Q&A] [answered]
First of all......"my hat's off to you sir...".  Incredible piece of python!  Well thought out...much better than my PERL scripts I  scrounged up.....and the Python version is extremely reliable.  I think I'll compile it and deploy when my debugging is done.
My setup:
Two 350's, Static IP's, RPi4B (staticIP), Your new Ver3.08 running beautifully.  Single config file, both boxes defined, URLid defined, Remotes fully operational.  No latency, Max resolution at 60FPS.  Pretty much plug and play with a couple of exceptions...
I did have to declare "python3" in the script to init in my env before it would run.  
My only issue....  The RPi boots, the service starts and shows running, detection of 350's fails.  The "slinger" log indicates two attempts to login to each defined 350, then states...."sorry I give up" (or something like that).  If I restart the service, after boot, all good.  Tried to add delays to the service, "wait on network" etc....to no avail.  Any ideas?

**Comment by GerryDazoo** [comment]
You didn't supply the logs. :-(.  So the first thing the code does it try to make a connection to the Slingboxes with the ip/port in the config file if it fails it tries to scan the local network for a box to use. But it finds two boxes and gives up because it can't decide which one to use. There's a thread for each box so the code may be working for just one of the two, it would be nice to see the logs...
So the logs will indicate which box/boxes it can't find on startup. Is it only one or is it both? The service should only startup after the network is up, so there is something else going on.

**Reply by anonyada** [reply]
Gerry,
Thx for getting back to me.....here's everything:
Please note: Running as user pi, or root, no difference.
By the way..."Debug" flag in config???  Syntax??
===========================================
******* CONFIG:
[SLINGBOXES]
sb1=350_1
sb2=350_2

;
[350_1]
sbtype="350/500/M1"
password=mypassval1
ipaddress=192.168.5.53
port=5201
; Resolution=12
Resolution=15
FrameRate=60
; VideoBandwidth=2000
VideoBandwidth=3000
VideoSmoothness=63
VideoSource=1
maxstreams=2
; StartChannel=355
;RemoteLock=yes
Remote=/home/pi/Slinger-main/CustomRemotes/remote_directv-h24.txt

[350_2]
sbtype="350/500/M1"
password=mypassval2.....
ipaddress=192.168.5.47
port=5201
; Resolution=12
Resolution=15
FrameRate=60
; VideoBandwidth=2000
VideoBandwidth=3000
VideoSmoothness=63
VideoSource=1
maxstreams=2
; StartChannel=355
;RemoteLock=yes
Remote=/home/pi/Slinger-main/CustomRemotes/remote_directv-h24.txt


[SERVER]
; local port number for the server to listen on for connections
; Send all streaming and Remote request to this port number.
port=8089
maxremotestreams=2
URLbase=mybaseurlval
enableremote=yes
Debug=yes



********Systemd service:
[Unit]
Description=SlingBox Slinger Server Service
After=multi-user.target

[Service]
Type=idle
User=root
WorkingDirectory=/home/pi/Slinger-main
StandardOutput=append:/tmp/sling.log
StandardError=append:/tmp/sling.log
ExecStart=/usr/bin/python3 -u slingbox_server.py 
Restart=always

[Install]
WantedBy=multi-user.target
----------------

******  LOG FOR NORMAL  systemd "enable/start"  (after system already booted):
--------------
Version : 3.08c Running on Linux-5.10.103-v7l+-armv7l-with-debian-10.13 pid= 2144
Using config file config.ini
Connection Manager Running on port 8089 with 2 max streams using URL grynchvid.
BOXES [('sb1', '350_1'), ('sb2', '350_2')]
Streamer Running: Streamer Running:   starting up on  port 808922   config.ini8388608config.ini 
 350_1350_2  350_1350_2  80898089  83886088388608

Checking for slingbox atChecking for slingbox at  ('192.168.5.47', 5201)('192.168.5.53', 5201)

('192.168.5.47', 5201) OK
('192.168.5.53', 5201)350_2  OKUsing slingbox at 
 ('192.168.5.47', 5201)
350_1Streamer:   Using slingbox at 350_2  ('192.168.5.53', 5201)Waiting for first stream, flushing any IR requests that arrive while not connected to slingbox

Streamer:  350_1 Waiting for first stream, flushing any IR requests that arrive while not connected to slingbox
Building page for 350_1
Reading Custom Remote definition from /home/pi/Slinger-main/CustomRemotes/remote_directv-h24.txt
Building page for 350_2
Reading Custom Remote definition from /home/pi/Slinger-main/CustomRemotes/remote_directv-h24.txt
 * Serving Flask app "slingbox_server" (lazy loading)
 * Environment: production
   WARNING: Do not use the development server in a production environment.
   Use a production WSGI server instead.
 * Debug mode: off
--------------------------------------
************FAILED COLD REBOOT RESULTS Log file:
Version : 3.08c Running on Linux-5.10.103-v7l+-armv7l-with-debian-10.13 pid= 697
Using config file config.ini
Connection Manager Running on port 8089 with 2 max streams using URL grynchvid.
BOXES [('sb1', '350_1'), ('sb2', '350_2')]
Streamer Running:  2 config.ini 350_1 350_1 8089 8388608
starting up on  port 8089 8388608
Checking for slingbox at ('192.168.5.53', 5201)
Streamer Running: Error connecting to   2('192.168.5.53', 5201) 
config.ini 350_2 350_2 8089 8388608
Checking for slingbox at ('192.168.5.47', 5201)
Error connecting to  ('192.168.5.47', 5201)
350_2 No valid slingbox ip info found in config.ini
Finding Slingbox on local network. My IP Info =  dc:a6:32:63:ee:d4350_1 
Finding Slingbox on local network. My IP Info =  dc:a6:32:63:ee:d6
Can't find a slingbox on network. Please make sure it's plugged in and connected. Check config.ini
No valid slingbox ip info found in config.ini
Finding Slingbox on local network. My IP Info =  dc:a6:32:63:ee:d4
Finding Slingbox on local network. My IP Info =  dc:a6:32:63:ee:d6
Can't find a slingbox on network. Please make sure it's plugged in and connected. Check config.ini
Building page for 350_1
Reading Custom Remote definition from /home/pi/Slinger-main/CustomRemotes/remote_directv-h24.txt
Building page for 350_2
Reading Custom Remote definition from /home/pi/Slinger-main/CustomRemotes/remote_directv-h24.txt
 * Serving Flask app "slingbox_server" (lazy loading)
 * Environment: production
   WARNING: Do not use the development server in a production environment.
   Use a production WSGI server instead.
 * Debug mode: off
350_2 Giving up. Sorry...
350_1 Giving up. Sorry...
---------------------------

**Reply by GerryDazoo** [reply]
Well it sure looks like the code can't reach the slingboxes on a reboot. 
Try putting this in your service definition
After=network-online.target
Wants=network-online.target

**Reply by anonyada** [reply]
No Joy.  (Tried that earlier....but did it again....).
Thought it might be DNS related...so I added those IP's to /etc/hosts.  Still no joy.
Interesting problem.....
Latest Logs after cold boot/ and service "restart":

Version : 3.08c Running on Linux-5.10.103-v7l+-armv7l-with-debian-10.13 pid= 717
Using config file 350multi_config.ini
Connection Manager Running on port 8089 with 2 max streams using URL grynchvid.
BOXES [('sb1', '350_1'), ('sb2', '350_2')]
Streamer Running:  2 350multi_config.ini 350_1 350_1 8089 8388608
Checking for slingbox atStreamer Running:  2  ('192.168.5.53', 5201)350multi_config.ini
 Error connecting to  ('192.168.5.53', 5201)
starting up on  port 8089 8388608
350_2 350_2 8089 8388608
Checking for slingbox at ('192.168.5.47', 5201)
Error connecting to  ('192.168.5.47', 5201)
350_1 No valid slingbox ip info found in config.ini
Finding Slingbox on local network. My IP Info =  dc:a6:32:63:ee:d4
Finding Slingbox on local network. My IP Info =  dc:a6:32:63:ee:d6
Can't find a slingbox on network. Please make sure it's plugged in and connected. Check config.ini
350_2 No valid slingbox ip info found in config.ini
Finding Slingbox on local network. My IP Info =  dc:a6:32:63:ee:d4
Finding Slingbox on local network. My IP Info =  dc:a6:32:63:ee:d6
Can't find a slingbox on network. Please make sure it's plugged in and connected. Check config.ini
Building page for 350_1
Reading Custom Remote definition from /home/pi/Slinger-main/CustomRemotes/remote_directv-h24.txt
Building page for 350_2
Reading Custom Remote definition from /home/pi/Slinger-main/CustomRemotes/remote_directv-h24.txt
 * Serving Flask app "slingbox_server" (lazy loading)
 * Environment: production
   WARNING: Do not use the development server in a production environment.
   Use a production WSGI server instead.
 * Debug mode: off
350_1 Giving up. Sorry...
350_2 Giving up. Sorry...
Version : 3.08c Running on Linux-5.10.103-v7l+-armv7l-with-debian-10.13 pid= 1176
Using config file 350multi_config.ini
Connection Manager Running on port 8089 with 2 max streams using URL grynchvid.
BOXES [('sb1', '350_1'), ('sb2', '350_2')]
Streamer Running:  Streamer Running: 2  350multi_config.ini 2starting up on  port 8089350_1  8388608
 350multi_config.ini350_1  350_28089  8388608350_2 
Checking for slingbox at8089 ('192.168.5.53', 5201) 
8388608
Checking for slingbox at('192.168.5.53', 5201)  OK('192.168.5.47', 5201)

350_1 ('192.168.5.47', 5201) Using slingbox at  OK('192.168.5.53', 5201)

350_2 Using slingbox at Streamer:   ('192.168.5.47', 5201)
350_1Streamer:   350_2Waiting for first stream, flushing any IR requests that arrive while not connected to slingbox 
Waiting for first stream, flushing any IR requests that arrive while not connected to slingbox
Building page for 350_1
Reading Custom Remote definition from /home/pi/Slinger-main/CustomRemotes/remote_directv-h24.txt
Building page for 350_2
Reading Custom Remote definition from /home/pi/Slinger-main/CustomRemotes/remote_directv-h24.txt
 * Serving Flask app "slingbox_server" (lazy loading)
 * Environment: production
   WARNING: Do not use the development server in a production environment.
   Use a production WSGI server instead.
 * Debug mode: off

**Reply by GerryDazoo** [reply]
Well, as you suggested maybe you can stick in a small delay to see if that helps. I didn't have this issue on my PI but I can't retest 'cus it's in a box 1000 miles away. So for now all I can suggect is putting in a short delay before the server starts.
I can't test this but sometime like  ExecStart="sleep 60;python.... blah blah blah"

**Reply by anonyada** [reply]
Thx....will do....going through the system logs now.....I'll eventually figure it out and post back.

**Reply by GerryDazoo** [reply]
Instead of just doing some dead reckoning on the time delay, you could turn the ExecStart into running a bash script where you could wait for the boxes to come online. Using either ping or nc (netcat). 
I'll noodle on adding some kind of retry/timeout on the initial "ping" of the slingbox.

**Reply by anonyada** [reply]
Gerry,
After reviewing the logs /var/log/syslog, it was painfully obvious that the "sling.service" had started way too soon,
long before DHCP had finished loading.  It's a RPi.  It's Debian....so yeah....we're beating our heads against the wall.

There are many different opinions on how to address the problem....all subject to network startup tags etc.  PIA!

My solution:
/usr/systemd/system/sling.service <Stop the service, and Disable the Service.  Keep the [UNITS] section your default.
(Don't delete the file, we still need it!)

Now Create file:
/usr/systemd/system/sling.timer  <--(Note name must match the .service file)
The contents:
# ----------------------
[Unit]
Description=Wait 60sec before starting the sling.service file

[Timer]
OnActiveSec=60sec
AccuracySec=1s

[Install]
WantedBy=timers.target
# ---------------------------
Enable it.  IE:  systemctl enable sling.timer.

sling.service (in it's original form) will now get called by sling.timer (because the name is the same and it's enabled.)
Works like a champ.
Hope this helps.

**Reply by GerryDazoo** [reply]
Great stuff, Thanks for looking into this.

**Reply by anonyada** [reply]
NP.  By the way....I ported my RPi4B image (Buster 32bit) to a RP3B+ (1Gb ram).  In case anyone's interested it works flawlessly.  Two slings, wifi 5G, online 24/7.  With both slings delivering HD video, the CPU usage is at 6%.  Perfect standalone solution.
***ALSO***  I have locked the IP's on my slings, and have blocked their WAN access.  They have officially disappeared on sling.com under my account.  You rock dude!

**Reply by DoogieHowitzerWMD** [reply]
>_**ALSO**_ I have locked the IP's on my slings, and have blocked their WAN access. They have officially disappeared on sling.com under my account. You rock dude!

What is involved in blocking off their WAN access? Having the boxes be invisible to Sling seems like a good precaution, but since I am running my Slinger server on an external network, I'm thinking I probably can't have one without the other.

**Reply by anonyada** [reply]
I run a ASUS WiFi 2.4/5G mesh.  Control of individual WAN access is available by MAC and IP.  While slings are DHCP configured by nature, I lock those individual IP's down to their MAC addresses respectively.  Therefore my slings always get the same IP address at boot.  I then lock those IP's for local LAN connectivity only.  They disappeared from the sling account immediately.    If you're running any kind of *NIX based server for your slings, you can do exactly the same thing with a subnet, local DHCP server, and IPTABLES.

**Reply by BanditRider2K** [reply]
This worked for me on a RPi 3B+ running Pi OS. Except systemd  is in /etc  not /usr.

**Reply by GerryDazoo** [reply]
The readme in the repository is correct. Can to ping who ever gave you this bogus info and have them correct it?

**Reply by BanditRider2K** [reply]
My point is sling.service needs to wait for the network to be ready before running.

**Reply by almoney** [reply]
I have a Pull Request pending that updates the service file to fix this.

This is what you need to change:

Description=SlingBox Server Service
delete: After=multi-user.target
insert:
After=systemd-hostnamed.service
Wants=systemd-hostnamed.service

**Reply by anonyada** [reply]
> This worked for me on a RPi 3B+ running Pi OS. Except systemd is in /etc not /usr.

That is correct.  However, the original manageable source resides in /usr/….
It’s ONLY after you correctly use the appropriate enable service command for your OS that the service is cataloged and copied to /etc directory.

**Reply by anonyada** [reply]
> I have a Pull Request pending that updates the service file to fix this.
> 
> This is what you need to change:
> 
> Description=SlingBox Server Service delete: After=multi-user.target insert: After=systemd-hostnamed.service Wants=systemd-hostnamed.service

This method has proved to be unreliable.  However if it works for you…fantastic! There’s nothing wrong with the original service description provided. It’s up to the end user to “season to taste” as needed IMHO.  I used a 60sec timer.  Past experience with Debian has revealed certain network updates over time can affect service start function.

**Comment by GerryDazoo** [comment, marked as answer]
PS. There is no debug flag for my code. That log is something that gets printed out by the web server code I've integrated (Flask).
